File: Main/static/JS/game/effects/toxicity.js
import { Modificator } from "../classes/modificator.js";
import { DurableEffect } from "./effect.js";

class Toxicity extends DurableEffect{
    constructor(hero, game_state, duration, caster, value, poison_duration, value_generator=null){
        super('ядовитость', 'toxicity.png', hero, game_state, duration, caster)
        this.value = value
        this.value_generator = value_generator
        this.poison_duration = poison_duration
    }

    get_value(){
        if (this.value_generator !== null) return this.value_generator(this)
        return this.value
    }

    gen_description(){
        return `Токсичность[${this.get_value()}-${this.poison_duration}] (длительность: ${this.duration})`
    }

    gen_modificators(){
        let value = this.value
        if (this.value_generator !== null){
            value = this.value_generator(this)
        }
        return [new Modificator('toxicity_value', 1, value, 'after', 'sum', ['autoattack_effects']),
                new Modificator('toxicity_duration', 1, this.poison_duration, 'after', 'sum', ['autoattack_effects'])]
    }

    after_move(){
        this.decrease_duration()
    }
}


export {Toxicity}

File: Main/models/effects/toxicity.py
from .effect import DurableEffect
from django.db import models
from Main.models.modifier import Modifier


class Toxicity(DurableEffect):
    name = 'Toxicity'
    img_src = 'toxicity.png'

    value = models.IntegerField()
    poison_duration = models.IntegerField()

    def gen_description(self, link):
        return f"Токсичность [Отравление: {self.value} | {self.poison_duration}] ({self.duration} ходов) (наложено {link.caster.name})"

    def after_move(self, link):
        self.decrease_duration(link)

    def gen_modifiers(self, link):
        return [Modifier(hero=link.hero, param_name='toxicity_value', value=self.value, duration=1),
                Modifier(hero=link.hero, param_name='toxicity_duration', value=self.poison_duration, duration=1)]
